cis194/week3/daniel-deng/src/Exercise02.hs

```diff
@@ -0,0 +1,11 @@
+module CIS194.Homework03.Exercise02 (localMaxima) where
+
+import Data.Maybe
+
+localMax :: (Integer, Integer, Integer) -> Maybe Integer
+localMax (x, y, z)
+  | y > x && y > z = Just y
+  | otherwise      = Nothing
+
+localMaxima :: [Integer] -> [Integer]
+localMaxima xs = mapMaybe localMax $ zip3 xs (drop 1 xs) (drop 2 xs)
```

Comment:
```haskell
window n = filter ((>= n) . length) . fmap (take n) . tails
```
Thanks, that's a really cool one. Question is though, what can we do with this windowed list of lists? Do we need to pattern match array in the helper function like this?

```haskell
localMax :: [Integer] -> Maybe Integer
localMax [x, y, z] | y > x && y > z = Just y
localMax _ = Nothing

localMaxima :: [Integer] -> [Integer]
localMaxima = mapMaybe localMax . window 3
```

In production code which one tends to be used more? The uncurried version (the tuple version) is clearer to me despite the fact that we still need guard, as the signature is more restrictive. What do you think? @shineli1984

Reply:
In production, we’ll use an advanced data type to embed length into List type. This is called dependent type(natural number dependent list type in this case). So that lists of different length are of different type.

In a less advanced code base, we’ll usually use a window function. But without the `filter ((>= n) . length)` part because the function receive its output will still need to match on all possible list length.

You are right that a triple is more expressive of the intend here. It’s a compromise compared to dependent type. And a list is arguably a little bit worse.

Reply:
@shineli1984 any reference to morediscussions on dependent type?